### PR TITLE
thread: add Windows implementation

### PR DESCRIPTION
**Pull request by sesmith177**
Signed-off-by: Sam Smith <sesmith177@gmail.com>

*Description*:
Adds a Win32 implementation of the ThreadImpl  class

*Risk Level*:
Low
*Testing*:
`bazel build //source/... && bazel test //test/...`. Also ran the tests on our Windows fork
*Docs Changes*:
N/A
*Release Notes*:
N/A

### CODE REVIEW

**Comment by zyfjeff** [comment]
Why not use std::thread?

**Comment by mattklein123** [comment]
> Why not use std::thread?

Historically, std::thread eats exceptions and prints useless crash/abort messages.

**Comment by mattklein123** [comment]
@jmarantz @curiouserrandy can you take a look to see how this might fit in with the thread factory stuff? Not sure if it matters or not.

**Comment by jmarantz** [comment]
@eziskind is working on the thread factory.

**Comment by curiouserrandy** [comment]
I don't think I have relevant expertise on this PR.  I'll unassign myself (if I can figure out how); please feel free to assign me or ping me if there's some reason I should be included.

**Comment by sesmith177** [comment]
@jmarantz not a problem, we'll wait for that to be merged and then rework this PR

**Comment by jmarantz** [comment]
This looks good so far; might get a little simpler if you are adjusting only the implementation of ThreadSystem; hopefully #5055 will get merged soon.

**Comment by jmarantz** [comment]
#5055 is now merged; and might make this change to support an alternate thread-system much more surgical.

Check out Thread::threadFactoryForTest() in test/test_common/utility.cc which you could probably just ifdef.

And check out the origin of the production thread-factory in source/exe/main_common.cc in the main_common() function. You could make your windows binary share MainCommonBase, but split off the main_common() function to another file which is not compiled in windows, and instantiate an alternate ThreadFactory in your new main().

File: source/common/thread/win32/thread_impl.cc [hunk 1]
```diff
@@ -0,0 +1,28 @@
+#include "common/common/assert.h"
+#include "common/thread/thread_impl.h"
+
+namespace Envoy {
+namespace Thread {
+
+ThreadImpl::ThreadImpl(std::function<void()> thread_routine) : thread_routine_(thread_routine) {
+  RELEASE_ASSERT(Logger::Registry::initialized(), "");
+  thread_handle_ = ::CreateThread(NULL, 0,
```
Comment (yuval-k) [inline source/common/thread/win32/thread_impl.cc]: drive-by question: should this be _beginthreadex?  (from the [docs](https://docs.microsoft.com/en-us/windows/desktop/api/processthreadsapi/nf-processthreadsapi-createthread): `A thread in an executable that calls the C run-time library (CRT) should use the _beginthreadex` )

**Comment by sesmith177** [comment]
@jmarantz I've updated this PR to take into account #5055 and to touch fewer files

@yuval-k You're right -- thanks for catching that

File: include/envoy/thread/thread.h [hunk 2]
```diff
@@ -18,6 +37,8 @@ class Thread {
    * Join on thread exit.
    */
   virtual void join() PURE;
+
+  virtual ThreadHandle handle() const PURE;
```
Comment (eziskind) [inline include/envoy/thread/thread.h]: Do we need to expose the thread handle? I don't see this method getting called anywhere.

Comment (sesmith177) [reply include/envoy/thread/thread.h]: We have some code in our fork (specifically the Windows version of Filesystem::WatcherImpl) that needs access to the thread handle. We haven't PR'd those changes back upstream yet but plan to eventually

Comment (jmarantz) [reply include/envoy/thread/thread.h]: For this PR can we remove the handle, as it's not used yet?

Then in the next PR we can discuss the handle and its usage in FilesystemImpl. I predict that I'm would suggest you make a windows-specific subclass of that too, but can't tell until I see it :)

In general I hope we can approach portability with injection of alternate class implementations rather than ifdefs, particularly ifdefs in an interface header.

Comment (sesmith177) [reply include/envoy/thread/thread.h]: Not a problem. 

The specific usage is our workaround for the fact that libevent can only watch sockets on Windows (as opposed to arbitrary). But we'll get to that in a future PR

File: source/common/common/win32/thread_impl.cc [hunk 3]
```diff
@@ -0,0 +1,31 @@
+#include <process.h>
+
+#include "common/common/assert.h"
+#include "common/common/thread_impl.h"
+
+namespace Envoy {
+namespace Thread {
+
+ThreadImpl::ThreadImpl(std::function<void()> thread_routine) : thread_routine_(thread_routine) {
```
Comment (jmarantz) [inline source/common/common/win32/thread_impl.cc]: It looks like this is the exact same name as the posix one, so the decision is made at link-time, and if you link both files in you'd get a duplicate symbol conflict.

That's not really a concern for windows vs posix, but you may also want an alternate thread system even if more than one is possible on the platform. That's something that we need (google has some custom thread infrastructure).

I'll give you another example: you may want a stats-collecting thread-system decorator that keeps a gauge of the active thread-count or other metrics.  I'd keep that separate from the platform-specific impls, so (e.g.) you could use that to decorate either windows or posix one.

Can we inject the right platform-specificThreadFactory where needed rather than having this be a link-time decision?

File: source/exe/main_common_base.cc [hunk 4]
```diff
@@ -125,42 +125,4 @@ void MainCommonBase::adminRequest(absl::string_view path_and_query, absl::string
...
-                               prod_component_factory_,
-                               std::make_unique<Runtime::RandomGeneratorImpl>(), thread_factory_);
-    return main_common.run() ? EXIT_SUCCESS : EXIT_FAILURE;
-  } catch (EnvoyException& e) {
-    return EXIT_FAILURE;
-  }
-}
-
-} // namespace Envoy
+} // namespace Envoy
```
Comment (jmarantz) [inline source/exe/main_common_base.cc]: add newline

File: test/test_common/utility.cc [hunk 5]
```diff
@@ -347,7 +347,11 @@ MockedTestAllocator::~MockedTestAllocator() {}
 namespace Thread {
 
 ThreadFactory& threadFactoryForTest() {
-  static ThreadFactoryImpl* thread_factory = new ThreadFactoryImpl();
+#if !defined(WIN32)
```
Comment (jmarantz) [inline test/test_common/utility.cc]: flip if/else and invert the conditional

File: source/exe/win32/main.cc [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+  }
+  WinsockCleanup winsock_cleanup;
+
+  std::unique_ptr<Envoy::MainCommonBase> main_common_base;
+  std::unique_ptr<Envoy::OptionsImpl> options;
+  Envoy::Event::RealTimeSystem real_time_system;
+  Envoy::DefaultTestHooks default_test_hooks;
+  Envoy::ProdComponentFactory prod_component_factory;
+  Envoy::Thread::ThreadFactoryImplWin32 thread_factory;
+  auto hotRestartVersion = [](uint64_t, uint64_t, bool) -> std::string { return "disabled"; };
```
Comment (jmarantz) [inline source/exe/win32/main.cc]: I think I like the way this is going, but there's pretty significant swaths of common code that is replicated. Would you consider factoring that out and putting it in main_common_base.cc?

File: include/envoy/thread/thread.h [hunk 7]
```diff
@@ -7,9 +7,22 @@
...
+// <windows.h> defines some macros that interfere with our code, so undef them
+#undef DELETE
+#undef GetMessage
+#endif
+
 namespace Envoy {
 namespace Thread {
 
+#if !defined(WIN32)
+using ThreadId = int32_t;
```
Comment (jmarantz) [inline include/envoy/thread/thread.h]: We can avoid ifdefs in the interface here and instead have ThreadId be an abstract base class, which I'd give two or three methods:
  -- compare to another ThreadID
  -- determine if it's the current thread
  -- serialize to a string for debug purposes

in the past I've done this this way: https://github.com/apache/incubator-pagespeed-mod/blob/20fbc5575c8fc825a137f6f1b2015829a9768f81/pagespeed/kernel/base/thread_system.h#L117

Note referenced there: https://linux.die.net/man/3/pthread_self
```
POSIX.1 allows an implementation wide freedom in choosing the type used to represent a
thread ID; for example, representation using either an arithmetic type or a structure is
permitted. Therefore, variables of type pthread_t can't portably be compared using the
C equality operator (==); use pthread_equal(3) instead.
```

File: source/common/common/posix/thread_impl.cc [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  RELEASE_ASSERT(rc == 0, "");
+}
+
+ThreadId currentThreadId() {
+#ifdef __linux__
+  return syscall(SYS_gettid);
+#elif defined(__APPLE__)
+  uint64_t tid;
+  pthread_threadid_np(NULL, &tid);
+  return static_cast<int32_t>(tid);
```
Comment (jmarantz) [inline source/common/common/posix/thread_impl.cc]: this is probably pre-existing, but IMO further evidence that I think we want an abstract base class for ThreadID.

I could see also a subclass for Apple and a subclass for Linux I guess.

File: source/common/common/posix/thread_impl.h [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+namespace Envoy {
+namespace Thread {
+
+/**
+ * Wrapper for a pthread thread. We don't use std::thread because it eats exceptions and leads to
+ * unusable stack traces.
+ */
+class ThreadImplPosix : public Thread {
+public:
+  ThreadImplPosix(std::function<void()> thread_routine);
```
Comment (jmarantz) [inline source/common/common/posix/thread_impl.h]: FWIW I suspect you don't need ThreadImplPosix to be exposed in the header -- it can be private to the ThreadFactoryImplPosix impl and in the cc file. This is not a big deal but I feel like it's nice to minimize what has to go into h files, for compile-speed and for human understanding of interfaces.

File: test/test_common/utility.cc (hunk 5 above)
Comment (sesmith177) [reply test/test_common/utility.cc]: So far, we've been following this pattern (`if !defined(WIN32)`) in our PRs. We don't have a strong opinion on which way the conditional should be written, but do think we should be consistent about which way we pick.

There's about ~8 other places we'd have to change as well, all but 1 in test code. If we want to make the change as part of this PR, that's fine with us.

File: source/common/common/posix/thread_impl.cc (hunk 8 above)
Comment (jmarantz) [reply source/common/common/posix/thread_impl.cc]: agreed; no need to add complexity here.

File: source/common/common/posix/thread_impl.h (hunk 9 above)
Comment (sesmith177) [reply source/common/common/posix/thread_impl.h]: I agree that we don't need ThreadImplPosix exposed in the header, but we do need ThreadImplWin32 exposed (so one of our Windows-specific classes can access data we don't want to put on the interface). It may be more consistent to have both exposed rather than one exposed and one placed in the cc file

**Comment by jmarantz** [comment]
ping when ready for another round.

File: source/common/common/posix/thread_impl.h (hunk 9 above)
Comment (jmarantz) [reply source/common/common/posix/thread_impl.h]: OK leave it exposed then, this is not a big deal, and if we do subclass Apple vs Linux in the future we'll need the shared impl exposed at least as a class declaration.

**Comment by sesmith177** [comment]
@jmarantz I think we're good to go for another round.

**Comment by jmarantz** [comment]
I see you made an update and I will review; but can't ATM; plan to get to it tonight.

File: include/envoy/thread/thread.h [hunk 10]
```diff
@@ -10,6 +10,17 @@
 namespace Envoy {
 namespace Thread {
 
+class ThreadId {
+public:
+  virtual ~ThreadId() {}
+
+  virtual std::string string() const PURE;
```
[the remark this answers is not in the document]
Comment (sesmith177) [reply include/envoy/thread/thread.h]: This gets used by the GuardDog to log an error in PANIC: https://github.com/envoyproxy/envoy/blob/43fc77901ee8a1608239daaa51b531f922628083/source/server/guarddog_impl.cc#L68

File: include/envoy/thread/thread.h [hunk 11]
```diff
@@ -10,6 +10,17 @@
...
+  virtual bool operator==(const ThreadId& rhs) const PURE;
```
[the remark this answers is not in the document]
Comment (sesmith177) [reply include/envoy/thread/thread.h]: I'm pretty sure this isn't used; we can remove it

File: include/envoy/thread/thread.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (jmarantz) [reply include/envoy/thread/thread.h]: I see...so the reason to call it 'string()', fmt::format will call it automatically without having to write out
```         PANIC(fmt::format("GuardDog: one thread ({}) stuck for more than watchdog_kill_timeout",
                          watched_dog.dog_->threadId()->debugString());
```
is that right? Do we do that elsewhere in Envoy? My personal preference is to have the explicit call but I'm OK with doing it this way if you put in a comment that's why it has to be called string().

Also I'm wondering if that works as you expect because threadId() actually returns a unique_ptr.

[the remark this answers is not in the document]
Comment (jmarantz) [reply include/envoy/thread/thread.h]: Oh I see...sorry I had confused a few different blocks of code, and I did comment on this above as well. WatchDog::threadId() returns a std::string and I think it'd be better to return a ThreadId and convert it if needed in the panic call.

File: source/exe/main_common_base.cc [hunk 12]
```diff
@@ -40,11 +40,9 @@ Runtime::LoaderPtr ProdComponentFactory::createRuntime(Server::Instance& server,
...
 }
 
-MainCommonBase::MainCommonBase(OptionsImpl& options, Event::TimeSystem& time_system,
-                               TestHooks& test_hooks, Server::ComponentFactory& component_factory,
-                               std::unique_ptr<Runtime::RandomGenerator>&& random_generator,
-                               Thread::ThreadFactory& thread_factory)
-    : options_(options), component_factory_(component_factory), thread_factory_(thread_factory) {
+MainCommonBase::MainCommonBase(OptionsImpl& options, Thread::ThreadFactory& thread_factory)
+    : options_(options), thread_factory_(thread_factory),
+      random_generator_(std::make_unique<Envoy::Runtime::RandomGeneratorImpl>()) {
```
[the remark this answers is not in the document]
Comment (sesmith177) [reply source/exe/main_common_base.cc]: The idea was that OptionsImpl and the ThreadFactory were the only parameters that were changing in the open-source code, and so we could remove duplication by having those as the only arguments to the constructor. We can revert this bit and deduplicate in a different manner

**Comment by jmarantz** [comment]
ping me when this is ready for another round

**Comment by sesmith177** [comment]
@jmarantz I think we've addressed everything except for some of the common code in main.cc -- did you have a specific idea about how to deduplicate without changing the MainCommonBase constructor?

File: source/exe/win32/main.cc (hunk 6 above)
Comment (jmarantz) [reply source/exe/win32/main.cc]: I still think we can factor out more, but I also understand the need to move this PR forward. Can we have a TODO here to factor out the common parts of this with the posix version? 

Maybe the right thing is to move all that into MainCommonBase as an alternate constructor. In any case, please don't iterate more on that now, let's follow up. There's tons of great stuff in here.

But I think you are going to have to do this refactoring rather than someone else, as no one else will be able to touch this code unless CI for Windows is added.

**Comment by jmarantz** [comment]
@htuch PTAL as this touches on the desire for the libevent evolution in #4952 and the proposed Api::Api refactor proposed in #5126.  Also see #3390.

File: source/common/common/posix/thread_impl.cc [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+#elif defined(__APPLE__)
+#include <pthread.h>
+#endif
+
+namespace Envoy {
+namespace Thread {
+
+namespace {
+
+int32_t getCurrentThreadId() {
```
[the remark this answers is not in the document]
Comment (htuch) [reply source/common/common/posix/thread_impl.cc]: Yeah, this seems a bit broken, @zuercher should we just fix this?

File: source/common/common/posix/thread_impl.h [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace Envoy {
+namespace Thread {
+
+class ThreadIdImplPosix : public ThreadId {
+public:
+  ThreadIdImplPosix(int32_t id);
+
+  std::string debugString() const override;
+
```
Comment (htuch) [inline source/common/common/posix/thread_impl.h]: Nit: prefer for the `override`s that there is no space between them and we start the block with something like `// ThreadId overrides.`

File: source/common/common/posix/thread_impl.h [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+/**
+ * Wrapper for a pthread thread. We don't use std::thread because it eats exceptions and leads to
+ * unusable stack traces.
+ */
+class ThreadImplPosix : public Thread {
+public:
+  ThreadImplPosix(std::function<void()> thread_routine);
+
+  /**
+   * Join on thread exit.
```
Comment (htuch) [inline source/common/common/posix/thread_impl.h]: Yeah, no need to repeat comments that are in the superclass, just a redirect to that is fine for overrides.

File: source/common/common/win32/thread_impl.h [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+ * unusable stack traces.
+ */
+class ThreadImplWin32 : public Thread {
+public:
+  ThreadImplWin32(std::function<void()> thread_routine);
+  ~ThreadImplWin32();
+
+  /**
+   * Join on thread exit.
+   */
```
Comment (htuch) [inline source/common/common/win32/thread_impl.h]: Same comments about comments in this file as with the POSIX thread implementation header.

File: source/exe/BUILD [hunk 17]
```diff
@@ -48,7 +48,10 @@ envoy_cc_library(
 
 envoy_cc_library(
     name = "envoy_main_entry_lib",
-    srcs = ["main.cc"],
+    srcs = select({
+        "//bazel:windows_x86_64": ["win32/main.cc"],
```
Comment (htuch) [inline source/exe/BUILD]: Can we do an `envoy_platform_deps` here and elsewhere? You might need to change the structure to have this as an explicit library.

File: source/exe/win32/main.cc (hunk 6 above)
Comment (htuch) [reply source/exe/win32/main.cc]: I think we should clean this up a bit more before shipping. Specifically, this PR introduces some duplication and a drift in behavior for common parts of main. See below for the exception handling, which is more fine grained here, but probably also should be in the POSIX version as it's not tied to platform.

I think we should just have some hooks for platform initialization that get called, and most of this file can be the same.

File: test/common/network/connection_impl_test.cc [hunk 18]
```diff
@@ -1563,7 +1567,10 @@ TEST_P(ReadBufferLimitTest, SomeLimit) {
 
 class TcpClientConnectionImplTest : public testing::TestWithParam<Address::IpVersion> {
 protected:
-  TcpClientConnectionImplTest() : dispatcher_(time_system_) {}
+  TcpClientConnectionImplTest()
+      : api_(Api::createApiForTest(stats_store_)), dispatcher_(time_system_, *api_) {}
+  Stats::IsolatedStoreImpl stats_store_;
```
Comment (htuch) [inline test/common/network/connection_impl_test.cc]: Nit: prefer a blank line between constructor/destructors and the member variables (here and elsewhere).

File: source/exe/win32/main.cc (hunk 6 above)
Comment (sesmith177) [reply source/exe/win32/main.cc]: I think the exception handling is the same in both versions -- see https://github.com/greenhouse-org/envoy/blob/37c183286611023e2fac690cb5f1d149d02a3aae/source/exe/posix/main.cc#L25-L33 for the POSIX version.

The exception handling is different if the `main_common` function: https://github.com/greenhouse-org/envoy/blob/37c183286611023e2fac690cb5f1d149d02a3aae/source/exe/posix/main_common.cc#L53 is used, though

**Comment by sesmith177** [comment]
@htuch @jmarantz I've given the deduplication of main.cc etc. another go

File: source/common/common/posix/thread_impl.cc (hunk 13 above)
[the remark this answers is not in the document]
Comment (zuercher) [reply source/common/common/posix/thread_impl.cc]: Sure.

File: include/envoy/thread/thread.h [hunk 19]
```diff
@@ -10,6 +10,16 @@
...
 
+class ThreadId {
+public:
+  virtual ~ThreadId() {}
+
+  virtual std::string debugString() const PURE;
+  virtual bool isCurrentThreadId() const PURE;
+};
+
+typedef std::unique_ptr<ThreadId> ThreadIdPtr;
```
[the remark this answers is not in the document]
Comment (jmarantz) [reply include/envoy/thread/thread.h]: I think we need to go with unique_ptr here (or shared_ptr) because the type can't be expressed concretely in the portable layer.  Is there a specific concern with unique_ptr here?

See http://man7.org/linux/man-pages/man3/pthread_self.3.html
```
POSIX.1 allows an implementation wide freedom in choosing the type used to
represent a thread ID; for example, representation using either an arithmetic
type or a structure is permitted. Therefore, variables of type pthread_t can't
portably be compared using the C equality operator (==); use pthread_equal(3)
instead.

Thread identifiers should be considered opaque: any attempt to use a
thread ID other than in pthreads calls is nonportable and can lead to
unspecified results.
```

[the remark this answers is not in the document]
Comment (htuch) [reply include/envoy/thread/thread.h]: That's fine. I just would skip a level of indirection if not needed.

File: source/exe/posix/legacy_main.h [hunk 20]
```diff
@@ -0,0 +1,18 @@
+#pragma once
+
+#include "server/options_impl.h"
+
+namespace Envoy {
+
+/**
+ * Legacy implementation of main_common.
```
Comment (htuch) [inline source/exe/posix/legacy_main.h]: @jmarantz should we just nuke this in this PR and have Envoy import on-call fix up the Google-side code?

File: source/common/common/posix/thread_impl.h [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+};
+
+/**
+ * Implementation of ThreadFactory
+ */
+class ThreadFactoryImplPosix : public ThreadFactory {
+public:
+  ThreadFactoryImplPosix() {}
+
+  ThreadPtr createThread(std::function<void()> thread_routine) override;
```
Comment (htuch) [inline source/common/common/posix/thread_impl.h]: Nit: can you fix the spacing/comments of these overrides (and other places that weren't updates to the `// <interface name>\n<dense block of override>` style below?

File: source/exe/win32/platform_setup.h [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+
+#include "absl/debugging/symbolize.h"
+
+// clang-format off
+#include <winsock2.h>
+// clang-format on
+
+namespace Envoy {
+
+class PlatformSetup {
```
Comment (htuch) [inline source/exe/win32/platform_setup.h]: Nit: classes should be named as nouns and methods as verbs, so maybe call this `PlatformMain` or something like that.

File: source/exe/posix/legacy_main.h (hunk 20 above)
Comment (jmarantz) [reply source/exe/posix/legacy_main.h]: Actually I think there are no callers in Google to main_common(). Let's just nuke. I can send a separate PR if that helps.

Comment (jmarantz) [reply source/exe/posix/legacy_main.h]: https://github.com/envoyproxy/envoy/pull/5298

File: source/exe/main_common.cc [hunk 23]
```diff
@@ -125,10 +125,10 @@ void MainCommonBase::adminRequest(absl::string_view path_and_query, absl::string
   });
 }
 
-MainCommon::MainCommon(int argc, const char* const* argv)
+MainCommon::MainCommon(int argc, const char* const* argv, PlatformMain& platform_main)
     : options_(argc, argv, &MainCommon::hotRestartVersion, spdlog::level::info),
       base_(options_, real_time_system_, default_test_hooks_, prod_component_factory_,
-            std::make_unique<Runtime::RandomGeneratorImpl>(), thread_factory_) {}
+            std::make_unique<Runtime::RandomGeneratorImpl>(), platform_main.threadFactory()) {}
```
Comment (jmarantz) [inline source/exe/main_common.cc]: Rather than changing the API to MainCommon why don't you just instantiate PlatformMain as a member var in MainCommon, at least for now. I think that would touch less code. Also this will avoid breaking our import :)

Can we rename PlatformMain to PlatformApi as there's nothing really 'Main' about it,  And actually it's taking on an overlapping role to Api::Api, bringing us back to to @htuch's https://github.com/envoyproxy/envoy/issues/5126 .

We should also either fix Api::createApiForTest in test/test_common/utility.cc to use PlatformApi, or put a TODO to do so.

Comment (sesmith177) [reply source/exe/main_common.cc]: The initial idea for this PlatformSetup / PlatformMain / PlatformApi object was:
1. constructor does platform specific initialization
2. member variables / accessors for any platform specific objects (currently ThreadFactory, we'll probably add some sort of filesystem object soon)
3. destructor does platform specific cleanup

As currently implemented, we don't want the destructor (and hence WSACleanup) to be called until we're completely done.  So constructing one to use in Api::createApiForTest probably won't work (at least on Windows).

Looking at this discussion (as well as the discussion in #5126) we may want to eventually pull out 2 into a different object from 1+3. (separate the platform api from the setup). That refactor feels a little large for this specific PR, though

File: test/test_common/utility.cc [hunk 24]
```diff
@@ -187,30 +187,27 @@ std::vector<std::string> TestUtility::split(const std::string& source, const std
...
+  ::mkdir(name.c_str(), S_IRWXU);
 #endif
 }
 
 void TestUtility::createSymlink(const std::string& target, const std::string& link) {
-#if !defined(WIN32)
-  const int rc = ::symlink(target.c_str(), link.c_str());
-  ASSERT_EQ(rc, 0);
-#else
+#if defined(WIN32)
```
Comment (jmarantz) [inline test/test_common/utility.cc]: I've seen you use `_MSC_VER` elsewhere and WIN32 here. Why's that? If possible can we use just one condition?

Also why `#if defined(WIN32)` rather than the more concise `#ifdef WIN32`?

Usually I see the `define(xxx)` form in the context of a more complex logical expression: `if defined(WIN32) || defined(APPLE)`

File: test/test_common/utility.cc [hunk 25]
```diff
@@ -187,30 +187,27 @@ std::vector<std::string> TestUtility::split(const std::string& source, const std
...
-  const int rc = ::rename(old_name.c_str(), new_name.c_str());
-  ASSERT_EQ(0, rc);
-#else
+#if defined(WIN32)
   // use MoveFileEx, since ::rename will not overwrite an existing file. See
   // https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/rename-wrename?view=vs-2017
   const BOOL rc = ::MoveFileEx(old_name.c_str(), new_name.c_str(), MOVEFILE_REPLACE_EXISTING);
   ASSERT_NE(0, rc);
+#else
+  const int rc = ::rename(old_name.c_str(), new_name.c_str());
```
Comment (jmarantz) [inline test/test_common/utility.cc]: I think all this should be inside the filesystem abstraction (include/envoy/filesystem/filesystem.h). Can you at least add a TODO for now?

Note that the Filesystem and ThreadFactory are both part of Api::Api so you can see the convergence of all this.

File: source/exe/main_common.cc (hunk 23 above)
Comment (jmarantz) [reply source/exe/main_common.cc]: I think this fits into createApiForTest but may require some auditing. The caller of that should ensure the returned Api::ApiPtr lives as long as needed.

Or is it a problem of some sort to have the ctor/dtor called on every unit test within a single process?

File: source/exe/posix/platform_main.h [hunk 26]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace Envoy {
+
+class PlatformMain {
+public:
+  PlatformMain(int, const char* const* argv) {
+#ifndef __APPLE__
+    // absl::Symbolize mostly works without this, but this improves corner case
+    // handling, such as running in a chroot jail.
+    absl::InitializeSymbolizer(argv[0]);
```
Comment (jmarantz) [inline source/exe/posix/platform_main.h]: should we have posix and windows implementions inherit from a common base-class that calls the symbolizer?

File: source/exe/main_common.cc (hunk 23 above)
Comment (jmarantz) [reply source/exe/main_common.cc]: Also one thing I've thought of in the past is that all Envoy tests that aren't purely portable code, could inherit from Envoy::Api::TestBase which instantiates the platform-specific API, providing access to thread systems, etc.

**Comment by jmarantz** [comment]
All my above suggeestions are for TODOs in this PR; the only thing I'd like to see changed in this PR is consistent use of a `#ifdef WIN32` if possible.

File: test/test_common/utility.cc (hunk 25 above)
Comment (sesmith177) [reply test/test_common/utility.cc]: We have this broken out here since we didn't want to add production code that is only called from tests. If moving this code into the filesystem abstraction is the way to go, we will do that as a part of the filesystem PR that we are currently working on.

File: source/exe/main_common.cc (hunk 23 above)
Comment (sesmith177) [reply source/exe/main_common.cc]: We also noticed that if we created a PlatformMain inside of createApiForTest and used that to give the Api a ThreadFactory reference, the reference would become invalid once the PlatformMain is destructed. So we would have to make sure the PlatformMain lasts as long as the Api object

**Comment by sesmith177** [comment]
We will replace `#if defined(XXX)` with `#ifdef XXX` where appropriate

File: source/exe/main_common.cc (hunk 23 above)
Comment (jmarantz) [reply source/exe/main_common.cc]: Yes, that's why I think it would be easier to either have envoy tests that require such objects inherit from a new Envoy::Api::TestBase, or they can simply old the ApiPtr in a test member variable. It obviously won't work to just do `createApiForTest().threadSystem()`, and this is why I suggested to make it a TODO.

Right now there's a code-smell of having the same ifdef in two places, and I don't think you need to fix it, but we should flag it for fixing.

Comment (sesmith177) [reply source/exe/main_common.cc]: Makes sense, we will add the TODO

File: source/exe/posix/platform_main.h (hunk 26 above)
Comment (jmarantz) [reply source/exe/posix/platform_main.h]: ...alternatively, you could just leave the call to InitializeSymbolizer in main() where it is now, and then you don't need to pass in argc,argv to PlatformMain.

**Comment by sesmith177** [comment]
@jmarantz added changes as requested

**Comment by jmarantz** [comment]
Hi -- I'm a relatively new maintainer and don't totally understand the flow yet. It looks like I can't merge this PR because there are still requested changes (although @htuch has LGTM'd so I think I should be able to merge).  See https://github.com/envoyproxy/envoy/pull/5072#pullrequestreview-184538542

It *looks* like you've addressed @htuch's suggestions but maybe you didn't click 'resolve conversation' for each comment?

**Comment by htuch** [comment]
@sesmith177 @jmarantz I'll take another pass now and merge if it looks all good!

File: source/exe/posix/platform_main.h [hunk 27]
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+#include "common/common/macros.h"
+#include "common/common/thread_impl.h"
+
+namespace Envoy {
+
+class PlatformMain {
```
Comment (htuch) [inline source/exe/posix/platform_main.h]: I'm still uncomfortable with this naming. `PlatformMain` to me should be a method, it's essentially an entry point, not a noun. Can we just make this `PlatformImpl`?

Comment (htuch) [reply source/exe/posix/platform_main.h]: LGTM sans this bikeshed :)

**Comment by sesmith177** [comment]
@htuch @lizan I've addressed the feedback

**Comment by htuch** [comment]
Needs master merge @sesmith177, will merge as soon as that happens.
